pkg/controller/controller.go

```diff
@@ -0,0 +1,388 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+// ResizeController watches PVCs and checks if they are requesting an resizing operation.
+// If requested, it will resize according PVs and update PVCs' status to reflect the new size.
+type ResizeController interface {
+	// Run starts the controller.
+	Run(workers int, leaderElectionConfig *util.LeaderElectionConfig)
+}
+
+type resizeController struct {
+	name            string
+	resizer         resizer.Resizer
+	kubeClient      kubernetes.Interface
+	claimQueue      workqueue.RateLimitingInterface
+	eventRecorder   record.EventRecorder
+	pvLister        corelisters.PersistentVolumeLister
+	pvSynced        cache.InformerSynced
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	pvcSynced       cache.InformerSynced
+	informerFactory informers.SharedInformerFactory
+}
+
+// NewResizeController returns a ResizeController.
+func NewResizeController(
+	name string,
+	resizer resizer.Resizer,
+	kubeClient kubernetes.Interface,
+	resyncPeriod time.Duration) ResizeController {
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
+	pvInformer := informerFactory.Core().V1().PersistentVolumes()
+	pvcInformer := informerFactory.Core().V1().PersistentVolumeClaims()
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events(v1.NamespaceAll)})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme,
+		v1.EventSource{Component: fmt.Sprintf("external-resizer %s", name)})
+
+	claimQueue := workqueue.NewNamedRateLimitingQueue(
+		workqueue.DefaultControllerRateLimiter(), fmt.Sprintf("%s-pvc", name))
+
+	ctrl := &resizeController{
+		name:            name,
+		resizer:         resizer,
+		kubeClient:      kubeClient,
+		pvLister:        pvInformer.Lister(),
+		pvSynced:        pvInformer.Informer().HasSynced,
+		pvcLister:       pvcInformer.Lister(),
+		pvcSynced:       pvcInformer.Informer().HasSynced,
+		claimQueue:      claimQueue,
+		eventRecorder:   eventRecorder,
+		informerFactory: informerFactory,
+	}
+
+	// Add a resync period as the PVC's request size can be resized again when we handling
+	// a previous resizing request of the same PVC.
+	pvcInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addPVC,
+		UpdateFunc: ctrl.updatePVC,
+		DeleteFunc: ctrl.deletePVC,
+	}, resyncPeriod)
+
+	return ctrl
+}
+
+func (ctrl *resizeController) addPVC(obj interface{}) {
+	objKey, err := getPVCKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
```

Comment:
hmm, do we really need to consider a PVC for resizing when it gets added/created for the first time?  I know it will be dropped by checks we have in place but it would cause unnecessary churn in controller loop.

Reply:
This is to deal with a rare situation:
1. `external-resizer` stopped, maybe stopped manually  by administrator, or just due to an unexpected panic
2. User edits a PVC and requests new size
3. `external-resizer` started.

After `external-resizer` started, it will only receive an `Added` event of this PVC from the `Informer`, without any `Updated` events. So if we just simply ignore the `Added` events, this resize request will lost.